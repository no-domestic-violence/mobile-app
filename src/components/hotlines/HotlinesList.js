import React, { useState, useEffect, useRef } from 'react';
import appApiClient from '../../api/appApiClient';
import {
  View,
  StyleSheet,
  FlatList,
  Linking,
  ActivityIndicator,
} from 'react-native';
import ListItem from './ListItem';
import { SearchBar } from 'react-native-elements';
export default function HotlinesList() {
  const [search, setSearch] = useState('');
  const [loading, setLoading] = useState(false);
  const [dataSource, setDataSource] = useState([]);
  const inputRef = useRef();

  useEffect(() => {
    setLoading(true);
    getHotlinesData();
  }, [search]);

  const makeCall = (phoneNumber) => {
    const iosPhoneNumber = `tel:${phoneNumber}`;
    Linking.openURL(iosPhoneNumber);
  };

  const getHotlinesData = async () => {
    try {
      const response = await appApiClient.get(`/hotlines`, {
        params: { searchTerm: search },
      });
      setDataSource([...response.data]);
      setLoading(false);
      inputRef.current.focus();
    } catch (error) {
      console.error(error);
    }
  };

  const ItemSeparatorView = () => {
    return <View style={styles.itemSeparator} />;
  };
  if (loading) {
    return (
      <View style={styles.container}>
        <ActivityIndicator size="large" color="#00ff00" />
      </View>
    );
  }
  return (
    <>
      {/* TODO: how to decode location  */}
      {/* <Text>Current Location</Text> */}
      <SearchBar
        ref={inputRef}
        inputStyle={{ backgroundColor: 'white' }}
        inputContainerStyle={{ backgroundColor: 'white' }}
        containerStyle={{ backgroundColor: 'white' }}
        autoCorrect={false}
        autoCapitalize="none"
        onChangeText={setSearch}
        placeholder="Type city or name"
        value={search}
      />

      <FlatList
        style={styles.list}
        data={dataSource}
        keyExtractor={(item, index) => index.toString()}
        ItemSeparatorComponent={ItemSeparatorView}
        enableEmptySections={true}
        renderItem={({ item }) => (
          <ListItem item={item} makeCall={makeCall} title={item.phone} />
        )}
      />
    </>
  );
}

const styles = StyleSheet.create({
  list: {
    flex: 1,
    padding: 10,
    color: 'black',
  },
  container: {
    flex: 1,
    justifyContent: 'center',
  },
  text: {
    color: 'black',
    fontFamily: 'Courier',
  },
  footer: {
    padding: 10,
    justifyContent: 'center',
    alignItems: 'center',
    flexDirection: 'row',
  },
  itemSeparator: {
    height: 0.5,
    width: '100%',
    backgroundColor: '#c8c8c8',
  },
});
