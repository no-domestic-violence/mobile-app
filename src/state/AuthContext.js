import AsyncStorage from '@react-native-async-storage/async-storage';
import createAppContext from './CreateAppContext';
import appApiClient from '../api/appApiClient';

const authReducer = (state, action) => {
  switch (action.type) {
    case 'SIGNUP_ERROR':
    case 'LOGIN_ERROR':
      return {
        ...state,
        errorMessage: action.payload,
      };
    case 'SIGNUP_SUCCESS':
    case 'LOGIN_SUCCESS':
      return {
        ...state,
        token: action.payload.token,
        username: action.payload.user.username,
        errorMessage: '',
      };
    case 'CHANGE_PASSWORD_SUCCESS':
      return {
        ...state,
        successMessage: action.payload,
      };
    case 'CHANGE_PASSWORD_ERROR':
      return {
        ...state,
        errorMessage: action.payload,
      };
    case 'AUTH_SUCCESS':
      return {
        ...state,
        token: action.payload.token,
        username: action.payload.username,
      };
    case 'LOGOUT':
      return {
        token: '',
        username: '',
      };
    case 'DELETE_ACCOUNT':
      return {
        token: '',
        username: '',
        successMessage: action.payload.data,
      };
    case 'REMOVE_ERRORS':
      return {
        ...state,
        errorMessage: '',
      };
    case 'REMOVE_MESSAGES':
      return {
        ...state,
        successMessage: '',
      };
    default:
      return state;
  }
};

const signup = (dispatch) => async ({ email, password, username }) => {
  try {
    const response = await appApiClient.post('/signup', {
      email,
      password,
      username,
    });

    await AsyncStorage.setItem('token', response.data.token);
    await AsyncStorage.setItem('username', response.data.user.username);

    dispatch({ type: 'SIGNUP_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'SIGNUP_ERROR',
      payload: "Something went wrong'( Try again",
    });
  }
};

const login = (dispatch) => async ({ email, password }) => {
  try {
    const response = await appApiClient.post('/login', { email, password });
    await AsyncStorage.setItem('token', response.data.token);
    await AsyncStorage.setItem('username', response.data.user.username);
    dispatch({ type: 'LOGIN_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'LOGIN_ERROR',
      payload: 'Are you sure about password and email?',
    });
  }
};

const authentication = (dispatch) => async () => {
  const token = await AsyncStorage.getItem('token');
  const username = await AsyncStorage.getItem('username');
  const user = { token, username };
  if (token) {
    dispatch({ type: 'AUTH_SUCCESS', payload: user });
  }
};
const removeErrors = (dispatch) => () => {
  dispatch({ type: 'REMOVE_ERRORS' });
};

const removeMessages = (dispatch) => () => {
  dispatch({ type: 'REMOVE_MESSAGES' });
};

const changePassword = (dispatch) => async ({
  email,
  password,
  oldPassword,
}) => {
  try {
    const response = await appApiClient.post('/changePassword', {
      email,
      oldPassword,
      password,
    });
    removeErrors();
    dispatch({ type: 'CHANGE_PASSWORD_SUCCESS', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'CHANGE_PASSWORD_ERROR',
      payload: error.response.data.message,
    });
  }
};

const signout = (dispatch) => async () => {
  await AsyncStorage.removeItem('token');
  dispatch({ type: 'LOGOUT' });
};

const deleteAccount = (dispatch) => async ({ username }) => {
  try {
    // TODO: fix sending params so it is more readable
    const response = await appApiClient.delete('/deleteUser', {
      params: { username },
    });
    await AsyncStorage.removeItem('token');
    await AsyncStorage.removeItem('username');
    dispatch({ type: 'DELETE_ACCOUNT', payload: response.data });
  } catch (error) {
    dispatch({
      type: 'CHANGE_PASSWORD_ERROR',
      payload: error.response.data.message,
    });
  }
};

export const { Provider, Context } = createAppContext(
  authReducer,
  {
    signup,
    login,
    signout,
    removeErrors,
    removeMessages,
    authentication,
    changePassword,
    deleteAccount,
  },
  { isLoggedIn: false, errorMessage: '' }
);
