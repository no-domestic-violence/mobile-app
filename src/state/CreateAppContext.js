import React, { useReducer } from 'react';

// boilerplate for creating Context and Provider
// does not to be changed
export default (reducer, actions, defaultValue) => {
  const Context = React.createContext();

  const Provider = ({ children }) => {
    const [state, dispatch] = useReducer(reducer, defaultValue);

    //actions is object for ex {updateCurrentLocation: (dispatch)=> (location) => { dispatch({ type: 'UPDATE_LOCATION', payload: location });}
    const boundActions = {};
    //we need to loop through each action in object and for every key(which is function) we call it with dispatch argument and it gives back function. this function we pass down to all children like ...boundActions
    for (const key in actions) {
      boundActions[key] = actions[key](dispatch);
    }
    return (
      <Context.Provider value={{ state, ...boundActions }}>
        {children}
      </Context.Provider>
    );
  };
  return { Context, Provider };
};
