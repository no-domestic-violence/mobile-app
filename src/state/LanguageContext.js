import React, { createContext, useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import i18n from 'services/i18n';
import PropTypes from 'prop-types';
import { DEFAULT_LANGUAGE } from '../services/i18n/language-detector';

const APP_LANGUAGE = 'lang';

const availableLanguages = [
  { lang: 'en', name: 'English' },
  { lang: 'de', name: 'German' },
];

export const LanguageContext = createContext({
  availableLanguages,
  appLanguage: DEFAULT_LANGUAGE,
  setAppLanguage: () => {},
  initializeAppLanguage: () => {},
});

export const LanguageProvider = ({ children }) => {
  const [appLanguage, setAppLanguage] = useState(DEFAULT_LANGUAGE);

  const setLanguage = async (language) => {
    i18n.changeLanguage(language);
    setAppLanguage(language);
    AsyncStorage.setItem(APP_LANGUAGE, language);
  };

  const initializeAppLanguage = async () => {
    const currentLanguage = await AsyncStorage.getItem(APP_LANGUAGE);
    if (!currentLanguage) {
      setLanguage(DEFAULT_LANGUAGE);
    } else {
      setLanguage(currentLanguage);
    }
  };

  return (
    <LanguageContext.Provider
      value={{
        availableLanguages,
        setAppLanguage: setLanguage,
        appLanguage,
        initializeAppLanguage,
      }}>
      {children}
    </LanguageContext.Provider>
  );
};
LanguageProvider.propTypes = {
  children: PropTypes.node.isRequired,
};
