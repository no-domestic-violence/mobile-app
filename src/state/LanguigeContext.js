//TODO: it did not work, to figure out if we need or we can just you provider for everything


// import React, { createContext } from 'react';
// import { I18nextProvider } from 'react-i18next';
// import i18next from 'i18next';

// const LanguigeContext = createContext();

// function LanguigeProvider(props) {

//   return (
//     <LanguigeContext.Provider>
//       <I18nextProvider i18n={i18next}>{props.children}</I18nextProvider>
//    </LanguigeContext.Provider>
//   );
// }

// export {LanguigeProvider, LanguigeContext};

