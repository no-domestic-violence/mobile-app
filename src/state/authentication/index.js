import createAppContext from 'helpers/createAppContext';
import {
  signup,
  login,
  signout,
  removeErrors,
  removeMessages,
  authentication,
  changePassword,
  deleteAccount,
  setAlreadyLaunchedValue,
} from './authentication.actions';
import authReducer from './authentication.reducer';

export const { Provider, Context } = createAppContext(
  authReducer,
  {
    signup,
    login,
    signout,
    removeErrors,
    removeMessages,
    authentication,
    changePassword,
    deleteAccount,
    setAlreadyLaunchedValue,
  },
  { isLoggedIn: false, errorMessage: '', isFirstLaunch: null }
);
