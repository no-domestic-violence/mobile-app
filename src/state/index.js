import React, {createContext, useState, useMemo} from 'react';

const AppContext = createContext();

function ContextProvider(props) {
  //just an example of context usage
  const [userToken, setUserToken] = useState('');
  const authContext = useMemo(() => {
    return {
      logIn: () => {
        setUserToken('tokenishere');
      },
      signUp: () => {
        setUserToken('tokenishere');
      },
      signOut: () => {
        setUserToken('');
      },
    };
  });

  return (
    <AppContext.Provider value={{authContext, userToken}}>
      {props.children}
    </AppContext.Provider>
  );
}

export {ContextProvider, AppContext};
