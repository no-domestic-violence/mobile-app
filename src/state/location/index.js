import createAppContext from 'helpers/createAppContext';

import locationReducer from './location.reducer';
import {
  updateCurrentLocation,
  fetchShelters,
  searchHotlinesByParam,
} from './location.actions';

export const { Provider, Context } = createAppContext(
  locationReducer,
  {
    updateCurrentLocation,
    fetchShelters,
    searchHotlinesByParam,
  },
  { currentLocation: null, shelters_list: [], hotlinesData: [] }
);
