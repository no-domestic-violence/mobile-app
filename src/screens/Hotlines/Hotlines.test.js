import React from 'react';
import { render } from '@testing-library/react-native';
import { LocationProvider } from 'state/';
import HotlinesList from './index';

describe('Hotlines screen', () => {
  const component = (
    <LocationProvider>
      <HotlinesList />
    </LocationProvider>
  );
  test('should have SearchBar and FlatList elements', async () => {
    const { getByTestId, getByPlaceholderText } = await render(component);
    expect(getByTestId('hotlinesFlatList')).not.toBeNull();
    expect(
      getByPlaceholderText(/Type city or organisation name/)
    ).not.toBeNull();
  });
  it('should match snapshot', () => {
    const result = render(component).toJSON();
    expect(result).toMatchSnapshot();
  });
});
