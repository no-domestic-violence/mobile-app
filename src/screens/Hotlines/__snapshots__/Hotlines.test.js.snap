// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hotlines screen should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#cadeee",
        "flexBasis": 0,
        "flexGrow": 1,
        "flexShrink": 1,
      },
      Object {
        "alignItems": "stretch",
        "flex": 1,
        "justifyContent": "center",
        "paddingHorizontal": 20,
      },
    ]
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#cadeee",
        "borderBottomColor": "#000",
        "borderBottomWidth": 1,
        "borderTopColor": "#000",
        "borderTopWidth": 0,
        "padding": 8,
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 0,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#cadeee",
            "borderBottomWidth": 0,
            "borderColor": "#86939e",
            "borderRadius": 3,
            "flexDirection": "row",
            "minHeight": 30,
            "overflow": "hidden",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
            "width": "90%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginLeft": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <
            icon={
              Object {
                "icon": Array [
                  512,
                  512,
                  Array [],
                  "f002",
                  "M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z",
                ],
                "iconName": "search",
                "prefix": "fas",
              }
            }
            size={20}
          />
        </View>
        <TextInput
          allowFontScaling={true}
          autoCapitalize="none"
          autoCorrect={false}
          editable={true}
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          placeholder="Type city or organisation name"
          placeholderTextColor="#86939e"
          platform="default"
          rejectResponderTermination={true}
          style={
            Object {
              "backgroundColor": "#cadeee",
              "color": "#86939e",
              "flex": 1,
              "fontSize": 18,
              "marginLeft": 10,
              "minHeight": 40,
              "width": "70%",
            }
          }
          testID="searchInput"
          underlineColorAndroid="transparent"
          value=""
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "height": 40,
              "justifyContent": "center",
              "marginRight": 8,
              "marginVertical": 4,
              "paddingRight": 4,
            }
          }
        >
          <View
            style={
              Object {
                "flexDirection": "row",
              }
            }
          />
        </View>
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "height": 0,
            "margin": 0,
            "padding": 0,
          }
        }
      />
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "#cadeee",
        "borderBottomColor": "#bcbbc1",
        "borderBottomWidth": 0.5,
        "height": 10,
      }
    }
  />
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    enableEmptySections={true}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "flex": 1,
      }
    }
    testID="hotlinesFlatList"
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
</View>
`;
